packages/asc-ui/src/components/Modal/Modal.tsx

```diff
@@ -30,34 +31,39 @@ const Modal: FunctionComponent<Props> = ({
   backdropOpacity,
   element,
   blurredNodeSelector,
-  className,
   onClose,
   ...otherProps
 }) => {
   const ref = useRef<HTMLDivElement>(null)
   const { keyDown } = useTrappedFocus(ref)
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const previouslyFocusedElement = useMemo(() => document.activeElement, [open])
 
   useEffect(() => {
-    let renderedTimer = 0
-    const { current: node } = ref
-    if (node) {
-      clearTimeout(renderedTimer)
-      renderedTimer = setTimeout(() => {
-        node.focus()
-      })
+    if (!open || !ref.current) {
+      return
     }
-    return () => {
-      clearTimeout(renderedTimer)
+
+    const firstFocusableEl = ref.current.querySelector(
+      focusableElements.join(', '),
+    )
+
+    if (firstFocusableEl instanceof HTMLElement) {
+      firstFocusableEl.focus()
     }
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [])
+  }, [open])
 
   const handleClose = () => {
     if (onClose) {
+      if (previouslyFocusedElement instanceof HTMLElement) {
+        previouslyFocusedElement.focus()
+      }
       onClose()
     }
```

Comment:
Can we add some tests to verify this re-focus logic is working correctly?

Reply:
Good call, I added a test